chanfitness: Rate limit in memory events based on peer flap rate

carlaKC opened the pull request:
This PR adds rate limiting to peers that we are currently tracking online/offline events for. This change prevents constantly flapping peers from filling up our memory with an endless set of online/offline events. It is preparatory work for #4165, addressing our in memory resource consumption before we proceed to write these events to disk. The rate limiting logic used is very simple, we just place peers in buckets based on the number of times they have flapped. This can easily be switched out in future, but we start simple. To add a tangible output to this PR, we display flap rate in listchannels. 

A new bucket is added to track peer flap count across restarts so that we maintain our rate limiting when lnd is restarted. 
This bucket will be extended to include uptime events.
```
[peers-bucket]
    [peer-pubkey]
        [flap-count-key]
       +[events-bucket]
```

This PR includes some refactoring to cleanup spacing/testing in the original implementation, and to update the system to store events per peer. While working on this, it became clear that having all of our peer level information stored in one place is a better design for our requirements (do not need to double store online events when we have two channels with a peer, flap count is consistently stored). If we need to extend this system to have channel level, we can still do so by having event logs for each channel within our peer log.

Crypt-iQ commented on chanfitness/chaneventstore.go:
```diff
@@ -41,80 +42,77 @@ type ChannelEventStore struct {
 // called.
-func NewChannelEventStore(config *Config) *ChannelEventStore {
+func NewChannelEventStore(config *Config,
+	now func() time.Time) *ChannelEventStore {
```
Why did you decide to not put the `now` func inside the config? I'm not opposed but I'm curious as to your reasoning. I think it's cleaner to modify the config but there might be other reasons to *not* put it in the config.

carlaKC replied:
At the time, it didn't feel like "config", but no strong reasoning. Moved into `Config` and gone for `clock.Clock` rather than just a now function, more flexible that way.

carlaKC commented on chanfitness/rate_limit.go:
```diff
@@ -0,0 +1,40 @@
+	// rateLimitScale is the number of events we allow per rate limited
+	// tier. Increasing this value makes our rate limiting more lenient,
+	// decreasing it makes us less lenient.
+	rateLimitScale = 200
```
Yeah, always difficult to hand wave a number for things like this. One thing I wanted to make sure of is that we don't punish peers that are restarting a "regular" number of times, because then we start getting worse uptime for legit peers. 

Did some rough calculations [here ](https://github.com/lightningnetwork/lnd/issues/4165#issuecomment-643252037) re what our worst case is for these values, but we can really plug anything in here, so open to suggestions!

Crypt-iQ replied:
Maybe let's see what others think, I'm fine with it

carlaKC commented on channeldb/db.go:
```diff
@@ -157,6 +157,12 @@ var (
+			// about our peers.
+			number:    17,
+			migration: mig.CreateTLB(peerBucket),
+		},
```
Yeah, we've started doing these for top level buckets. Nice because we don't have to always do nil-checks and create them on the fly. The last few migrations have been this type.

carlaKC commented on chanfitness/chaneventstore.go:
```diff
@@ -415,3 +470,18 @@ func (c *ChannelEventStore) GetUptime(channelPoint wire.OutPoint, startTime,
+	log.Debugf("recording flap count for: %v channels", len(c.channels))
+
+	for _, channel := range c.channels {
+		err := c.cfg.WriteFlapCount(channel.peer, channel.flapCount)
```
Indeed, nice suggestion!

carlaKC commented on chanfitness/chaneventstore.go:
```diff
@@ -346,6 +395,12 @@ func (c *ChannelEventStore) consume(subscriptions *subscriptions) {
 
+		case <-c.cfg.FlapCountTicker.Ticks():
+			if err := c.recordFlapCount(); err != nil {
+				log.Errorf("could not record flap "+
```
Should also be error 👍

carlaKC commented on chanfitness/chanevent.go:
```diff
@@ -90,11 +90,12 @@ type chanEventLog struct {
 // newEventLog creates an event log for a channel with the openedAt time set.
 func newEventLog(channelPoint wire.OutPoint, peer route.Vertex,
-	now func() time.Time) *chanEventLog {
+	now func() time.Time, flapCount uint32) *chanEventLog {
```
Going to leave for now, with the promise that my future-self will do so if we need to add more 😬

carlaKC commented on chanfitness/chanevent.go:
```diff
@@ -79,24 +107,66 @@ func (e *chanEventLog) close() {
+// including events that are staged for addition to our set of events but have
+// not yet been committed to (because we rate limit and store only the aggregate
+// outcome over a period).
+func (e *chanEventLog) listEvents() []*channelEvent {
```
Indeed, this would draw the line between event log and store more cleanly in the unit tests, which is nice. Got a few PRs coming up in this area so will keep it in mind!

Crypt-iQ replied:
👍

bhandras commented on chanfitness/rate_limit.go:
```diff
@@ -0,0 +1,40 @@
+// getRateLimit returns the value of the rate limited tier that we are on based
+// on current flap count. If a peer's flap count exceeds the top tier, we just
+// return our highest tier.
+func getRateLimit(flapCount uint32) time.Duration {
```
does `flapcount`being `uint32` add much? I'd just use an `int` to avoid all these casts. (Afaik Golang uses 64 bit ints in 64 bit archs, but I could be wrong)

bhandras commented on chanfitness/chaneventstore.go:
```diff
@@ -266,8 +307,19 @@ type subscriptions struct {
+	// the case of a graceful shutdown. If we do not shutdown gracefully,
+	// our worst case is data from our last flap count tick (1H).
+	defer func() {
+		defer c.wg.Done()
```
why not simply call at the end?

bhandras commented on chanfitness/chaneventstore.go:
```diff
@@ -414,3 +472,33 @@ func (c *ChannelEventStore) GetUptime(channelPoint wire.OutPoint, startTime,
+
+	flapCounts := make(map[route.Vertex]map[wire.OutPoint]uint32)
+
+	for _, channel := range c.channels {
```
is this goroutine `c.channels`' only accessor ?

carlaKC replied:
Yes, this whole subsystem is driven by the `consume` loop, so we don't need to worry about access.

bhandras commented on channeldb/peers_test.go:
```diff
@@ -0,0 +1,41 @@
+
+	var flapCount uint32 = 20
+	peers := map[route.Vertex]map[wire.OutPoint]uint32{
+		testPub: {
```
Test could have at least two peers and at least one peer could have more than one channel to make sure serialization is correct.

bhandras commented on channeldb/peers.go:
```diff
@@ -0,0 +1,158 @@
+	//     	|		|--flap-count-key: <flap count>
+	//      |
+	//      |-- <peer-pubkey>
+	//      |        |--<chan-point>
```
Something I don't understand is why do we store the flap count for each channel? Isn't it supposedly the same?

Crypt-iQ replied:
Ah I think I was unclear in my review comment a ways above. Perhaps a better architecture would be a sub-bucket under the peer bucket that stores the total count and then each chan point has its individual flap count. Reason being that granularity will help with banning (per-peer OR per-channel): https://github.com/lightningnetwork/lnd/issues/1183

carlaKC commented on chanfitness/chanevent.go:
```diff
@@ -5,6 +5,7 @@ import (
 	"time"
 
 	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/clock"
```
nice to be able to use it!

yyforyongyu commented on chanfitness/chaneventstore_test.go:
```diff
@@ -63,163 +63,164 @@ func TestStartStoreError(t *testing.T) {
-						IdentityPub:     pubKey,
-					},
-				}
+			generateEvents: func(ctx *chanEventStoreTestCtx) (wire.OutPoint,
```
nit: line too long

yyforyongyu commented on chanfitness/chaneventstore_test.go:
```diff
@@ -63,163 +63,164 @@ func TestStartStoreError(t *testing.T) {
-					},
-				}
+			generateEvents: func(ctx *chanEventStoreTestCtx) (wire.OutPoint,
+				route.Vertex) {
```
nit: missing new line

yyforyongyu commented on chanfitness/chaneventstore_test.go:
```diff
@@ -63,163 +63,164 @@ func TestStartStoreError(t *testing.T) {
-						IdentityPub:     pubKey,
-					},
-				}
+			generateEvents: func(ctx *chanEventStoreTestCtx) (wire.OutPoint,
```
nit: line too long

yyforyongyu commented on chanfitness/chaneventstore_test.go:
```diff
@@ -63,163 +63,164 @@ func TestStartStoreError(t *testing.T) {
-						IdentityPub:     pubKey,
-					},
-				}
+			generateEvents: func(ctx *chanEventStoreTestCtx) (wire.OutPoint,
```
nit: line too long

yyforyongyu commented on chanfitness/chaneventstore_test.go:
```diff
@@ -63,163 +63,164 @@ func TestStartStoreError(t *testing.T) {
-					},
-				}
+			name: "Channel opened, peer flaps, closed",
+			generateEvents: func(ctx *chanEventStoreTestCtx) (wire.OutPoint,
```
nit: line too long

yyforyongyu commented on chanfitness/chaneventstore_test.go:
```diff
@@ -63,163 +63,164 @@ func TestStartStoreError(t *testing.T) {
-						IdentityPub:     pubKey,
-					},
-				}
+			generateEvents: func(ctx *chanEventStoreTestCtx) (wire.OutPoint,
```
nit: line too long

yyforyongyu commented on chanfitness/chaneventstore_testctx_test.go:
```diff
@@ -0,0 +1,338 @@
+// assertPeerInfo asserts that we have tracked our peers online state and uptime
+// as expected. This function can only be called once the test context's store
+// has been stopped.
+func (c *chanEventStoreTestCtx) assertPeerInfo(expected map[route.Vertex]peerInfo) {
```
nit: line too long

carlaKC replied:
Thanks for pointing these out! In some cases, such as this one and a few `generateEvents` functions above, I've intentionally chosen to not line wrap. In the contributor guidelines, we make a note that wrapping at 80 should be done on a [best effort basis](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#48-additional-style-constraints-on-top-of-gofmt).

When we have a single long parameter for a function, I find it more readable to run over than add the newline (or make the var name less descriptive):
```
func (c *chanEventStoreTestCtx) assertPeerInfo(
    expected map[route.Vertex]peerInfo) {
```

This is a matter of personal taste, so there can always be arguments in either direction. Thankfully we have `gofmt` to save us time discussing these things in most cases!

yyforyongyu replied:
Cool thanks for letting me know!

carlaKC replied:
Thanks for reviewing! Always really helpful to have extra eyes :D

yyforyongyu commented on chanfitness/chaneventstore.go:
```diff
@@ -256,143 +325,179 @@ type subscriptions struct {
+			// A new channel has been opened, we must add the
+			// channel to the store and record a channel open event.
 			case channelnotifier.OpenChannelEvent:
+				compressed := event.Channel.IdentityPub.SerializeCompressed()
```
nit: line too long

yyforyongyu commented on chanfitness/chaneventstore.go:
```diff
@@ -256,143 +325,179 @@ type subscriptions struct {
+}
+
+// GetChanInfo gets all the information we have on a channel in the event store.
+func (c *ChannelEventStore) GetChanInfo(channelPoint wire.OutPoint) (*ChannelInfo,
```
nit: line too long

yyforyongyu commented on chanfitness/chaneventstore_test.go:
```diff
@@ -428,41 +367,44 @@ func TestGetUptime(t *testing.T) {
 			if test.channelFound {
-				store.channels[test.channelPoint] = &chanEventLog{
+				peer := route.Vertex{1}
+				ctx.store.channels[test.channelPoint] = &chanEventLog{
```
nit: line too long

yyforyongyu commented on chanfitness/chaneventstore_test.go:
```diff
@@ -63,163 +63,164 @@ func TestStartStoreError(t *testing.T) {
 				SubscribeChannelEvents: test.ChannelEvents,
 				SubscribePeerEvents:    test.PeerEvents,
 				GetOpenChannels:        test.GetChannels,
+				Clock:                  clock.NewTestClock(testNow),
```
nit: line too long

carlaKC commented on chanfitness/chaneventstore_testctx_test.go:
```diff
@@ -0,0 +1,236 @@
+
+// Quit returns the mock's quit channel. We close this when we stop the mock.
+func (m *mockSubscription) Quit() <-chan struct{} {
+	return m.quit
```
nice idea, added an assertion that the updates channel is closed. Ended up removing this function entirely and embedding the interface, since it's not used in the current set of tests.

bhandras commented on chanfitness/chaneventstore.go:
```diff
@@ -352,65 +301,72 @@ func (c *ChannelEventStore) consume(subscriptions *subscriptions) {
+
+	// Uptime is the total amount of time that the channel peer has been
+	// observed as online during the monitored lifespan.
+	Uptime time.Duration
```
Same question in a different form, but if a peer reportedly comes online such that it's always in the idle timeout then what is the uptime?

carlaKC replied:
So this would be the case where a peer connects to us, does not respond to any messages for the idle timeout, we disconnect and then they immediately reconnect but don't respond to any messages again?

With the current notifications, this peer would be up for that period of time, since we've defined online as having a connection open with that peer (so far). I think this case would likely be due to a mis-implemented peer, and while somebody could setup their node to "game" our uptime metric, the flap count would reflect that they go down every 5 minutes, and I don't think there's much to gain because they would have to respond to our messages to participate in forwards (the only incentive I can see for wanting high uptime, we choose to route through you).  

We could look at moving notifications for uptime tracking to a lower level, but I think we will then start running into ordering issues.

bhandras commented on chanfitness/chanevent.go:
```diff
@@ -29,74 +28,152 @@ func (e eventType) String() string {
-func (e *chanEventLog) close() {
-	e.closedAt = e.clock.Now()
+// addOnlineEvent records an online event in our event log.
+func (p *peerLog) addOnlineEvent(online bool, time time.Time) {
```
nit: a bit confusing that `addOnlineEvent` records both on and offline events, maybe consider changing the godoc to reflect this.

carlaKC replied:
updated to `addEvent` and updated doc, since we also have an `onlineEvent` function, the two can be confusing.

bhandras commented on chanfitness/interface.go:
```diff
@@ -0,0 +1,29 @@
+	onlineEvent(online bool)
+
+	// addChannel adds a new channel.
+	addChanel(channelPoint wire.OutPoint) error
```
nit: addChan**n**el

bhandras commented on chanfitness/chaneventstore.go:
```diff
@@ -268,7 +259,19 @@ func (c *ChannelEventStore) consume(subscriptions *subscriptions) {
 			// from the store and record a channel closed event.
 			case channelnotifier.ClosedChannelEvent:
-				c.closeChannel(event.CloseSummary.ChanPoint)
+				compressed := event.CloseSummary.RemotePub.SerializeCompressed()
```
Question (not related to this line): since in `consume()` we read from many channels, could we run into event ordering issues?

bhandras commented on chanfitness/chanevent.go:
```diff
@@ -95,14 +121,34 @@ func (p *peerLog) addOnlineEvent(online bool, time time.Time) {
 	eventType := peerOnlineEvent
 	if !online {
 		eventType = peerOfflineEvent
+		p.flapCount++
```
Adding to the prev comments: we could increase flap count for subsequent online events too.

carlaKC replied:
One thing we may need to account for if we increment for online events is that we will then do so on restart for every peer (not just those which are offline when we restart). Perhaps this is more fair, because the online peers may have flapped while our node was down, and the offline ones just have bad luck.

bhandras commented on channeldb/peers.go:
```diff
@@ -0,0 +1,99 @@
+	//      |
+	//      |-- <peer-pubkey>
+	//      |        |--flap-count-key: <flap count>
+	peerBucket = []byte("peers-bucket")
```
nit: could be `peersBucket` instead

bhandras commented on channeldb/peers.go:
```diff
@@ -0,0 +1,99 @@
+
+		flapBytes := peerBucket.Get(flapCountKey)
+		if flapBytes == nil {
+			return ErrNoFlapCount
```
Is this error needed? Do we expect flap count to be always serialized?

carlaKC replied:
Previous versions needed this in case we tried to lookup a peer that hadn't flushed to disk, but no longer needed after the refactor in chanfitness. Removed.

carlaKC commented on chanfitness/chanevent.go:
```diff
@@ -128,51 +134,72 @@ func (e *chanEventLog) getOnlinePeriods() []*onlinePeriod {
-			offline = false
+			// If our previous event is nil, we just set it and
+			// break out of the switch.
+			if lastEvent == nil {
```
I’ve had another look at this, and I think that we may be ok in terms of ordering because we notify on quite a high level. We notify peer online events in `addPeer` and offline events in `removePeer`. 

`addPeer` is only called when we get an inbound connection that we do not have an existing record of, or we replace a stale connection with a fresh one (`InboundPeerConnected`/ `OutboundPeerConnected`). When we are replacing stale connections, we remove all previous state, which will trigger an offline event before our new online event. 

This happens at a much higher level than our connection retries, and has logic in place that’s aware of existing connections. This does make our online events less granular, but it happens at the same level for every peer (and is preexisting for this uptime measure).  

However, it’s still difficult to say for sure whether we can rely on in-order events in reality. Perhaps @Crypt-iQ  can weigh in on this? I still like the suggestion of counting online and offline events, because it makes sure that our measurement isn’t biased (perhaps certain peers have certain odd connection habits, others don’t). Esp in this case, where the value itself matters less than the comparison of different peers values.

bhandras commented on chanfitness/chanevent.go:
```diff
@@ -212,7 +267,7 @@ type onlinePeriod struct {
 func (p *peerLog) getOnlinePeriods() []*onlinePeriod {
 	// Return early if there are no events, there are no online periods.
-	if len(p.onlineEvents) == 0 {
+	if len(p.listEvents()) == 0 {
```
nit: could save the result of `listEvents` to a variable to avoid creating the list twice.

bhandras commented on chanfitness/chaneventstore.go:
```diff
@@ -448,14 +459,20 @@ func (c *ChannelEventStore) getChanInfo(req channelInfoRequest) (*ChannelInfo,
 		flapCounts[peer] = &channeldb.FlapCount{
-			Count:    uint32(flapCount),
+			Count: uint32(
+				cooldownFlapCount(now, flapCount, *lastFlap),
```
iiuc this won't really cooldown the flapcount as we always get it from the peer monitor where the flapcount always increases so the max we could get is 0.9 * flapcount

carlaKC replied:
Moved this up and tested with some regtest nodes. Since we only need the cooldown for rate limiting, decided to keep it as we need it as you suggest. 

It is the case that we may cooldown then increment, but that makes sense to me in the following case:
- peer with 100 flap count connects, time since flap > cooldown period
- cooldown to 95
- update to 96, since they have just flapped, also update their last flapped

bhandras commented on Dockerfile:
```diff
@@ -1,43 +1,36 @@
-FROM golang:1.14.5-alpine as builder
+FROM golang:1.13-alpine as builder
```
Consider puting the `Dockerfile` changes in a separate commit.

bhandras replied:
But also, why did you reduce the version of the golang alpine container?

carlaKC replied:
rebase mistake, will remove this!

bhandras reviewed (changes requested):
I think the cooldown needs fixing, other than that, the PR looks good! :+1:

carlaKC commented:
Changes since last review [here](https://github.com/lightningnetwork/lnd/compare/423dd2b3167c031a1a336647f2a4ac2fc54715db..ca19aba3569dcc3e8c06ec37a996d8c208c6907a), other pushes are rebases.

bhandras commented on chanfitness/rate_limit.go:
```diff
@@ -35,3 +49,27 @@ func getRateLimit(flapCount int) time.Duration {
+	// Calculate time since our last flap, and the number of times we need
+	// to apply our cooldown factor.
+	timeSinceFlap := now.Sub(lastFlap)
+	cooldownPeriods := timeSinceFlap / flapCountCooldownPeriod
```
This way a flap of 100 would cool down to 1 in about 90 weeks.
We could be faster after the initial period. Also we could be more gradual if you allow fractional `coolDownPeriods` (just make `timeSinceFlap` float64).

bhandras commented on chanfitness/chaneventstore.go:
```diff
@@ -74,8 +81,22 @@ type Config struct {
 
+// peerMap is the map used to map peers to flap counts, declared here to allow
+// shorter function signatures.
+type peerMap map[route.Vertex]*channeldb.FlapCount
```
nit: wdyt about `peerFlapCountMap`?

bhandras reviewed (commented):
Almost ready, just one question left on my end.

bhandras reviewed (approved):
With the last changes, the PR LGTM, great work @carlaKC :100:

Crypt-iQ reviewed (approved):
Awesome work! Really like the usage of mock clock + deterministic unit tests. Tested this out locally with a flapping peer and could see the `listpeers` output increment and noticed no `pprof` usage for `chanfitness`. One thing is that both sides will have this flap count even though one of them disconnected first, that could be something addressed in a later PR, but I honestly don't know that it's necessary.
